File: app/src/main/java/com/huanchengfly/tieba/post/fragments/SearchForumFragment.java
package com.huanchengfly.tieba.post.fragments;


import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.huanchengfly.tieba.api.TiebaApi;
import com.huanchengfly.tieba.api.models.SearchForumBean;
import com.huanchengfly.tieba.post.R;
import com.huanchengfly.tieba.post.adapters.SearchForumAdapter;
import com.huanchengfly.tieba.post.components.MyLinearLayoutManager;
import com.huanchengfly.tieba.post.components.dividers.SearchDivider;
import com.huanchengfly.tieba.post.utils.ThemeUtil;

import org.jetbrains.annotations.NotNull;

import butterknife.BindView;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SearchForumFragment extends BaseFragment {
    public static final String TAG = "SearchForumFragment";

    public static final String ARG_KEYWORD = "keyword";
    @BindView(R.id.fragment_search_refresh_layout)
    SwipeRefreshLayout refreshLayout;
    @BindView(R.id.fragment_search_recycler_view)
    RecyclerView recyclerView;
    private String keyword;
    private SearchForumAdapter mAdapter;

    private SearchForumBean.DataBean mData;

    public SearchForumFragment() {
    }

    public static SearchForumFragment newInstance(String keyword) {
        SearchForumFragment forumFragment = new SearchForumFragment();
        Bundle bundle = new Bundle();
        bundle.putString(ARG_KEYWORD, keyword);
        forumFragment.setArguments(bundle);
        return forumFragment;
    }

    public void setKeyword(String keyword, boolean refresh) {
        this.keyword = keyword;
        if (refresh) {
            refresh();
        } else {
            this.mData = null;
            mAdapter.reset();
        }
    }

    @Override
    protected void onFragmentVisibleChange(boolean isVisible) {
        if (mData == null && isVisible) {
            refresh();
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            keyword = getArguments().getString(ARG_KEYWORD);
        }
    }

    @Override
    int getLayoutId() {
        return R.layout.fragment_search;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        recyclerView.setLayoutManager(new MyLinearLayoutManager(getAttachContext()));
        recyclerView.addItemDecoration(new SearchDivider(getAttachContext()));
        mAdapter = new SearchForumAdapter(getAttachContext());
        mAdapter.setLoadEndView(R.layout.layout_footer_loadend);
        mAdapter.setLoadFailedView(R.layout.layout_footer_load_failed);
        recyclerView.setAdapter(mAdapter);
        refreshLayout.setOnRefreshListener(this::refresh);
        ThemeUtil.setThemeForSwipeRefreshLayout(refreshLayout);
    }

    private void setRefreshing(boolean refreshing) {
        if (refreshLayout != null) refreshLayout.setRefreshing(refreshing);
    }

    private void refresh() {
        setRefreshing(true);
        TiebaApi.getInstance().searchForum(keyword).enqueue(new Callback<SearchForumBean>() {
            @Override
            public void onResponse(@NotNull Call<SearchForumBean> call, @NotNull Response<SearchForumBean> response) {
                mData = response.body().getData();
                mAdapter.setData(mData);
                setRefreshing(false);
                mAdapter.loadEnd();
            }

            @Override
            public void onFailure(@NotNull Call<SearchForumBean> call, @NotNull Throwable t) {
                setRefreshing(false);
                Toast.makeText(getAttachContext(), t.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    protected void onFragmentFirstVisible() {
        refresh();
    }
}

File: app/src/main/java/com/huanchengfly/tieba/post/fragments/SearchForumFragment.kt
package com.huanchengfly.tieba.post.fragments

import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import com.alibaba.android.vlayout.DelegateAdapter
import com.alibaba.android.vlayout.VirtualLayoutManager
import com.huanchengfly.tieba.post.R
import com.huanchengfly.tieba.post.activities.ForumActivity
import com.huanchengfly.tieba.post.adapters.HeaderDelegateAdapter
import com.huanchengfly.tieba.post.adapters.SearchForumAdapter
import com.huanchengfly.tieba.post.adapters.base.OnItemClickListener
import com.huanchengfly.tieba.post.api.TiebaApi.getInstance
import com.huanchengfly.tieba.post.api.models.SearchForumBean
import com.huanchengfly.tieba.post.components.MyViewHolder
import com.huanchengfly.tieba.post.dpToPx
import com.huanchengfly.tieba.post.interfaces.ISearchFragment
import com.huanchengfly.tieba.post.toastShort
import com.huanchengfly.tieba.post.utils.ThemeUtil
import com.scwang.smart.refresh.layout.SmartRefreshLayout
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class SearchForumFragment : BaseFragment(), ISearchFragment, OnItemClickListener<SearchForumBean.ForumInfoBean> {
    @JvmField
    @BindView(R.id.fragment_search_refresh)
    var refreshLayout: SmartRefreshLayout? = null

    @BindView(R.id.fragment_search_recycler_view)
    lateinit var recyclerView: RecyclerView

    private var keyword: String? = null
    private val layoutManager: VirtualLayoutManager by lazy { VirtualLayoutManager(attachContext) }
    private val delegateAdapter: DelegateAdapter by lazy { DelegateAdapter(layoutManager) }
    private val exactMatchAdapter: SearchForumAdapter by lazy {
        SearchForumAdapter(attachContext).apply {
            setOnItemClickListener(this@SearchForumFragment)
        }
    }
    private val fuzzyMatchAdapter: SearchForumAdapter by lazy {
        SearchForumAdapter(attachContext).apply {
            setOnItemClickListener(this@SearchForumFragment)
        }
    }
    private var mData: SearchForumBean.DataBean? = null

    override fun setKeyword(
            keyword: String?,
            needRefresh: Boolean
    ) {
        this.keyword = keyword
        if (needRefresh) {
            refreshLayout?.autoRefresh()
        } else {
            mData = null
            delegateAdapter.clear()
        }
    }

    override fun onFragmentVisibleChange(isVisible: Boolean) {
        if (mData == null && isVisible) {
            refreshLayout?.autoRefresh()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            keyword = requireArguments().getString(ARG_KEYWORD)
        }
    }

    public override fun getLayoutId(): Int {
        return R.layout.fragment_search
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        recyclerView.layoutManager = layoutManager
        recyclerView.adapter = delegateAdapter
        refreshLayout!!.setOnRefreshListener { refresh() }
        ThemeUtil.setThemeForSmartRefreshLayout(refreshLayout)
    }

    private fun refresh() {
        if (keyword == null) {
            return
        }
        getInstance().searchForum(keyword!!).enqueue(object : Callback<SearchForumBean> {
            override fun onResponse(call: Call<SearchForumBean>, response: Response<SearchForumBean>) {
                mData = response.body()!!.data
                refreshLayout?.finishRefreshWithNoMoreData()
                reloadAdapters()
            }

            override fun onFailure(call: Call<SearchForumBean>, t: Throwable) {
                t.message?.let { attachContext.toastShort(it) }
                refreshLayout?.finishRefresh(false)
            }
        })
    }

    private fun reloadAdapters() {
        delegateAdapter.clear()
        if (mData != null) {
            if (mData!!.exactMatch != null) {
                exactMatchAdapter.setData(listOf(mData!!.exactMatch!!))
                delegateAdapter.addAdapter(HeaderDelegateAdapter(
                        attachContext,
                        R.string.title_exact_match,
                        R.drawable.ic_round_graphic_eq
                ).apply {
                    setHeaderBackgroundResource(R.drawable.bg_top_radius_8dp)
                    topMargin = attachContext.resources.getDimensionPixelSize(R.dimen.card_margin)
                    startPadding = 16.dpToPx()
                    endPadding = 16.dpToPx()
                })
                delegateAdapter.addAdapter(exactMatchAdapter)
            }
            if (!mData!!.fuzzyMatch.isNullOrEmpty()) {
                fuzzyMatchAdapter.setData(mData!!.fuzzyMatch!!)
                delegateAdapter.addAdapter(HeaderDelegateAdapter(
                        attachContext,
                        R.string.title_fuzzy_match,
                        R.drawable.ic_infinite
                ).apply {
                    setHeaderBackgroundResource(R.drawable.bg_top_radius_8dp)
                    topMargin = attachContext.resources.getDimensionPixelSize(R.dimen.card_margin)
                    startPadding = 16.dpToPx()
                    endPadding = 16.dpToPx()
                })
                delegateAdapter.addAdapter(fuzzyMatchAdapter)
            }
        }
        delegateAdapter.notifyDataSetChanged()
    }

    override fun onFragmentFirstVisible() {
        refreshLayout!!.autoRefresh()
    }

    companion object {
        const val TAG = "SearchForumFragment"
        const val ARG_KEYWORD = "keyword"

        @JvmStatic
        @JvmOverloads
        fun newInstance(keyword: String? = null): SearchForumFragment {
            val forumFragment = SearchForumFragment()
            val bundle = Bundle()
            bundle.putString(ARG_KEYWORD, keyword)
            forumFragment.arguments = bundle
            return forumFragment
        }
    }

    override fun onClick(viewHolder: MyViewHolder, item: SearchForumBean.ForumInfoBean, position: Int) {
        item.forumName?.let { ForumActivity.launch(attachContext, it) }
    }
}